### PR TITLE
Use in-memory buffers instead of WASI for I/O

### PR DESCRIPTION
**Pull request by jeffcharles**
Replace using WASI's stdio with in-memory buffers for I/O. An exported function named `initialize` that takes the length of the input provides the address of the input buffer so the host can copy in data and an exported function named `finalize` returns a series of pointers and lengths for the output and logs.

This change both moves us off a standard which is slated for deprecation and also improves performance in terms of fuel usage.

This introduces a ~27% reduction in fuel (14,828 to 10,839) in one of the benchmark cases when excluding the initialization and finalization invocations from the fuel measurements.

### CODE REVIEW

File: provider/src/lib.rs [hunk 1]
```diff
@@ -30,12 +32,18 @@ thread_local! {
...
+    static OUTPUT_AND_LOG_PTRS: RefCell<[usize; 4]> = RefCell::new([0; 4]);
+}
+
 impl Default for Context {
     fn default() -> Self {
         Self {
             bump_allocator: Bump::new(),
             input_bytes: Vec::new(),
             output_bytes: ByteBuf::new(),
+            logs: Vec::with_capacity(1024),
```
Comment (saulecabrera) [inline provider/src/lib.rs:46]: Given that we have now landed in explicit `initialize` / `finalize` function invocations, I wonder if we should make this configurable from the host?

Comment (saulecabrera) [reply provider/src/lib.rs:46]: I think we briefly discuss that earlier on, maybe?

**Review by saulecabrera — COMMENTED** [review]
This is looking really good!

File: provider/src/lib.rs (hunk 1 above)
Comment (jeffcharles) [reply provider/src/lib.rs:46]: I can pass it as an additional parameter in initialize. But also, it's a vector so it'll grow beyond the initial capacity if it needs to. But using the default capacity definitely hurts performance.

Comment (jeffcharles) [reply provider/src/lib.rs:46]: Actually I guess another tradeoff is if the initial capacity changes, then the fuel consumption for a Function that performs logging may change as well. If we reduce the initial capacity, then a Function that performs enough logging to exceed the new capacity that would not have exceeded the old capacity would see an increase in fuel consumption. And if the initial capacity were increased, then a Function that performed enough logging to exceed the old capacity but not enough to exceed the new capacity would see a decrease in fuel consumption.

File: provider/src/log.rs [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+        assert_eq!(logs.len, 0);
+        assert_eq!(logs.offset, 0);
+        assert_eq!(ptr1, logs.buffer.as_ptr());
+        assert_eq!(len1, 0);
+        assert_eq!(len2, 0);
+        assert!(ptr2.is_null());
+    }
+
+    #[test]
+    fn test_append_exact_capacity() {
```
Comment (otheriault) [inline provider/src/log.rs:153]: Nice, I was curious about this edge case.

File: provider/src/write.rs [hunk 3]
```diff
@@ -205,30 +204,6 @@ decorate_for_target! {
...
-            }
-            if stdout.flush().is_err() {
-                return WriteResult::IoError;
-            }
-            WriteResult::Ok
-        })
-    }
-}
-
-#[cfg(not(target_family = "wasm"))]
```
Comment (otheriault) [inline provider/src/write.rs:231]: Should this be removed here?

Comment (jeffcharles) [reply provider/src/write.rs:231]: 👍 I'll re-add the attribute. Technically it doesn't change the provider's Wasm API since this function would require another attribute to make it an exported function when compiling to Wasm but nothing inside this crate needs this function so may as well not compile it in.
